### PR TITLE
Add the 'makeimage' service

### PR DESCRIPTION
**Pull request by catalin-manciu**
This adds a service for generating chart images based on the latest
benchmark results using the matplotlib module.
The URL for this service is /sample_project/makeimage/.
The image can be directly embedded in any HTML document using
 or simply accessed
using the same URL.

The mandatory GET params are:
env -	the environment's name
proj -	the project's name
branch -the branch's name
exe -	the executable's name
ben -	the benchmark's name
The following params are optional:
revs -	positive integer, number of results the image will display.
	revs=10 is the implicit value representing that the latest 10
	results will be displayed on the chart
relative - if '1' or 'yes', instead of displaying absolute values,
	the first value from the series is used as a baseline and
	the chart will display percentual increments from that baseline
base_commit - specifies a commit name to be used as a baseline instead
	of the first value of the series
base_env - the environment's name to be used for the base_commit instead
	of env
base_proj - the project's name to be used for the base_commit instead
	of proj
base_exe - the executable's name to be use for the base_commit instead of
	exe

### CODE REVIEW

File: codespeed/views.py [hunk 1]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+        if key not in data:
+            return 'Key "' + key + '" missing from GET request', True
+        elif data[key] == '':
+            return 'Value for key "' + key + '" empty in GET request', True
+
+    # Check that 'revs' is the correct format (if it exists)
+    if 'revs' in data:
+        try:
+            rev_value = int(data['revs'])
+        except:
```
Comment (tobami) [inline codespeed/views.py]: I'd prefer not to catch naked exceptions and instead catch `TypeError`

File: codespeed/views.py [hunk 2]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
 
     return HttpResponse("All result data saved successfully", status=202)
+
+
+def django_has_content_type():
+    return (django.VERSION[0] > 1 or
+            (django.VERSION[0] == 1 and django.VERSION[1] >= 6))
+
+
+def validate_results_request(data):
```
Comment (tobami) [inline codespeed/views.py]: This function could go into a separate `validators.py` module.

File: codespeed/views.py [hunk 3]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+        except:
+            return 'Value for key "revs" is not an integer', True
+        if rev_value <= 0:
+            return 'Value for key "revs" should be a strictly positive '\
+                   'integer', True
+
+    return '', False
+
+
+def get_benchmark_results(data):
```
Comment (tobami) [inline codespeed/views.py]: This could could go in `views_data.py`, where the other functions to fetch data for the views are.

File: codespeed/views.py [hunk 4]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+                            'Baseline project "' + data['base_proj'] +
+                            '" does not exist!')
+        except Executable.DoesNotExist:
+            return None, HttpResponseNotFound(
+                            'Baseline executable "' + data['base_exe'] +
+                            '"does not exist!')
+        except Branch.DoesNotExist:
+            return None, HttpResponseNotFound(
+                            'Baseline branch "' + data['base_branch'] +
+                            '" does not exist!')
```
Comment (tobami) [inline codespeed/views.py]: Instead of returning `None, err` all the time, we could return just the data or raise a django ObjectDoesNotExist exception (with the same message being set now). The use of the function (the view) could then catch the error and return the HttpResponseNotFound.

File: codespeed/views.py [hunk 5]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+            (django.VERSION[0] == 1 and django.VERSION[1] >= 6))
+
+
+def validate_results_request(data):
+    """
+    Validates that a result request dictionary has all needed parameters
+
+    It returns a tuple
+        "", False  when no errors where found
+        Error_message, True when there is an error
```
Comment (tobami) [inline codespeed/views.py]: I would change the return value to True, and raise a `ValidationError` (from django.core.exceptions) when there is an error.

File: codespeed/views.py [hunk 6]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+
+    err_msg, err = validate_results_request(data)
+
+    if err:
+        return HttpResponseBadRequest(err_msg)
+
+    result_data, error = get_benchmark_results(data)
+
+    if error is not None:
+        return error
```
Comment (tobami) [inline codespeed/views.py]: With the changes suggested this would be:

```python
try:
    validate_results_request(data)
except ValidationError as e:
    return HttpResponseBadRequest(str(e))

try:
     result_data = get_benchmark_results(data)
except ObjectDoesNotExist as e:
    return HttpResponseNotFound(str(e))
```

File: codespeed/views.py [hunk 7]
```diff
@@ -731,3 +739,239 @@ def add_json_results(request):
...
+
+    canvas_width = int(data['width']) if 'width' in data else 600
```
Comment (tobami) [inline codespeed/views.py]: All the image generation code could also be separated into a `images` module

File: codespeed/views.py (hunk 2 above)
Comment (catalin-manciu) [inline codespeed/views.py]: Moved validate_results_request to validators.py but there is still another validation function that should follow the same template (raising a ValidationError on error instead of returning a message, bool tuple): def validate_result(item) from results.py. I could move that to validators.py and maybe propose a more generic way to do validation in another pull request.

File: codespeed/views.py (hunk 5 above)
Comment (catalin-manciu) [inline codespeed/views.py]: Returning anything from this function didn't make much sense when signaling errors by throwing exceptions so, in the newest revision, this doesn't return anything).

File: codespeed/views.py (hunk 2 above)
Comment (tobami) [inline codespeed/views.py]: Good catch! `validate_result` has a bad interface. It returns `Environment, error`, so it really does two things: validate the input and fetch an environment. That tells us it should be split.

Anyway your suggestion makes total sense

File: codespeed/views.py (hunk 4 above)
Comment (tobami) [inline codespeed/views.py]: Yep, much cleaner now!

**Comment by tobami** [comment]
OK, so the build fails for Python 3 because of the `cStringIO` import. To make it work it would need to be changed to:

```python
try:
    from io import StringIO
except ImportError:
    from cStringIO import StringIO
```
(or possibly `BytesIO`.

If you change that, we should be able to merge!

**Comment by tobami** [comment]
@catalin-manciu Great, thanks for addressing the comments!
